### PR TITLE
[Plugin] Implement plugin getters for track and vehicle subpositions

### PR DESCRIPTION
**Pull request by Basssiiie**
Hey all,

I've been experimenting with the track iterator implemented in #16975 and I've been trying to use it to calculate `trackProgress`-distance between two vehicles. For this I need to know how long track pieces can be, and while there is [`TrackSegment.length`](https://github.com/OpenRCT2/OpenRCT2/blob/121afbe1d0eb1ff526b223468fc887bf8919e954/distribution/openrct2.d.ts#L1241), this is just an arbitrary number for each track piece and not an exact upper bound for `trackProgress`.

Thus, hereby a PR that exposes this upper bound, the subposition coordinates, and a new property for `Car` to get which subposition it is currently using.

- I wasn't sure about exposing `vehicle_get_move_info_size()` in the header file, but it seemed better than copy/pasting the contents. But if you have a better suggestion, I'd love to hear it.
- The `TrackSubposition` properties `angle` and `banking` are called that way to be consistent with the properties already present in `TrackSegment`.
- I also changed the type of `CoordsXYZD.direction` to be more compatible with all other uses of `Direction`. (This affects just the Typescript side, and thus is not a breaking change.)

Thank you for your time!

### CODE REVIEW

**Comment by frutiemax** [comment]
Related to #16964

**Comment by Basssiiie** [comment]
@frutiemax ah I hadn't seen that one. Would that PR have any effect on the external Javascript API? I see the different subposition numbers are still a thing, right? Or are there plans to deprecate the subposition type numbers any time in the future?

**Comment by frutiemax** [comment]
I don't think #16964 has any impact on the javascript api tbh, I thought this PR would use the arrays with the subpositions but it isn't.

**Comment by Basssiiie** [comment]
@frutiemax Would there be problems IF subpositions would be implemented in a similar fashion?

I'm trying to avoid issues where parts of the plugin API need to be deprecated again because of known future plans.

**Comment by frutiemax** [comment]
I don't really know about that, maybe @Gymnasiast can answer this?

**Comment by Gymnasiast** [comment]
No, I’m afraid I don’t know a great deal about that area.

**Comment by Basssiiie** [comment]
I ended up implementing the subposition coordinates API myself as well. :)

File: src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp [hunk 1]
```diff
@@ -141,4 +144,25 @@ DukValue ScTrackSegment::elements_get() const
     return DukValue::take_from_stack(ctx);
 }
 
+uint16_t ScTrackSegment::getSubpositionLength(uint8_t trackSubposition, uint8_t direction) const
+{
+    return vehicle_get_move_info_size(static_cast<VehicleTrackSubposition>(trackSubposition), _type, direction);
```
Comment (duncanspumpkin) [inline src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:149]: why would you need this you are getting a vector of the whole thing so you don't need this.

File: src/openrct2/scripting/bindings/ride/ScTrackSegment.h [hunk 2]
```diff
@@ -19,6 +19,18 @@
 
 namespace OpenRCT2::Scripting
 {
+    template<> inline DukValue ToDuk(duk_context* ctx, const rct_vehicle_info& value)
+    {
+        DukObject dukSubposition(ctx);
+        dukSubposition.Set("x", value.x);
+        dukSubposition.Set("y", value.y);
+        dukSubposition.Set("z", value.z);
+        dukSubposition.Set("direction", value.direction);
```
Comment (duncanspumpkin) [inline src/openrct2/scripting/bindings/ride/ScTrackSegment.h:28]: I would prefer it expose it as yaw, pitch and roll as that is the names ideally the codebase should be using.

File: src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp (hunk 1 above)
Comment (Basssiiie) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:149]: Because I think getting the whole vector may be less performant if you don't need the whole vector. (Since it's also creating a copy of all that data) The implementations for [tile elements](https://github.com/OpenRCT2/OpenRCT2/blob/7f29e4e39c76dbf7039162368ec8bb604e943d03/distribution/openrct2.d.ts#L792-L795) and [rides & entities](https://github.com/OpenRCT2/OpenRCT2/blob/7f29e4e39c76dbf7039162368ec8bb604e943d03/distribution/openrct2.d.ts#L616-L623) have the same functionality.

File: src/openrct2/scripting/bindings/ride/ScTrackSegment.h (hunk 2 above)
Comment (Basssiiie) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.h:28]: I can rename them but, as mentioned in the main post, it would add inconsistency with the already existing external plugin API use of 'angle', 'slope' and 'banking' [here](https://github.com/OpenRCT2/OpenRCT2/blob/7f29e4e39c76dbf7039162368ec8bb604e943d03/distribution/openrct2.d.ts#L1194-L1202). Is that inconsistency preferred over the current one?

Comment (duncanspumpkin) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.h:28]: those are for tracks though its a different incomparable property this is for vehicles.

File: src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp (hunk 1 above)
Comment (duncanspumpkin) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:149]: Would anything ever need that though? rides / entities is makes sense as you often want to quickly know how many rides exist. Why would you care about the length of the subposition array and why would you query it often?

Comment (Basssiiie) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:149]: It's the use case for why I made the PR to begin with:
> I've been trying to use it to calculate trackProgress-distance between two vehicles. For this I need to know how long track pieces can be, and while there is [`TrackSegment.length`](https://github.com/OpenRCT2/OpenRCT2/blob/121afbe1d0eb1ff526b223468fc887bf8919e954/distribution/openrct2.d.ts#L1241), this is just an arbitrary number for each track piece and not an exact upper bound for `trackProgress`.

File: src/openrct2/scripting/bindings/ride/ScTrackSegment.h (hunk 2 above)
Comment (Basssiiie) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.h:28]: In the end they're the same slope/banking/pitching values. :P But alright, I'll rename them.

File: src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp (hunk 1 above)
Comment (IntelOrca) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:149]: There is no API for getting a single sub position though, so how are you avoiding getting the whole vector?

Comment (Basssiiie) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:149]: @IntelOrca for just distance between two vehicles progress-wise you don't really need the actual positions. If car 1 is at trackprogress 5 out of 32 for a flat track piece, and car 2 is at trackprogress 12 out of 32 for the flat track piece after it, then their trackprogress distance is (32 - 12) + 5 = 25 between car 2 and car 1. For this calculation I don't need the subposition, but I do need the length of the vector.

In short, it's distance in progress, not xyz based distance.

Comment (Basssiiie) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:149]: But feel free to suggest ideas for better API endpoints!

Comment (IntelOrca) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:149]: Oh, so length does not mean the number of elements in an array, it is an actual metric within the game?

Comment (Basssiiie) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:149]: @IntelOrca The current `TrackSegment.length` draws its numbers from [this table](https://github.com/OpenRCT2/OpenRCT2/blob/a68f8e84678f1f92c7a6c1bf9410c96bb1e0d00d/src/openrct2/ride/TrackData.cpp#L2820) which is only used for calculating the ride length statistic. While some lengths there are the same as the subposition lengths (flat straight pieces), others are not (like turns).

Comment (Basssiiie) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:149]: For `getSubpositionLength` it does mean the array length. I added it to prevent pulling in the whole subposition vector if only the length is needed, similar to the API's for [tile elements](https://github.com/OpenRCT2/OpenRCT2/blob/7f29e4e39c76dbf7039162368ec8bb604e943d03/distribution/openrct2.d.ts#L792-L795) and [rides & entities](https://github.com/OpenRCT2/OpenRCT2/blob/7f29e4e39c76dbf7039162368ec8bb604e943d03/distribution/openrct2.d.ts#L616-L623).

But if it is preferred to have it removed, let me know.

Comment (IntelOrca) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:149]: Can you send me a snippet of a script where you use just the array length?

Comment (Basssiiie) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:149]: @IntelOrca Here's a Typescript snippet that gets the distance between a car and its preceding car (up to 5 track pieces away, if it has any).

```ts
const MaxForwardIterations = 5;


export function getSpacingToPrecedingVehicle(car: Car): number | null
{
	const precedingCarId = car.previousCarOnRide;
	if (precedingCarId === null)
	{
		return null;
	}

	const carInFront = <Car>map.getEntity(precedingCarId);

	const currentProgress = car.trackProgress;
	const inFrontProgress = carInFront.trackProgress;
	const currentTrackLocation = car.trackLocation;
	const inFrontTrackLocation = carInFront.trackLocation;

	if (isLocationEqual(currentTrackLocation, inFrontTrackLocation))
	{
		// Both cars are on the same track piece
		return (inFrontProgress - currentProgress);
	}

	const iterator = getTrackIteratorAtLocation(currentTrackLocation);
	if (!iterator)
	{
		return null;
	}

	let iteratorSegment = iterator.segment;
	if (!iteratorSegment)
	{
		console.log(`Track iterator does not have segment for car at position; ${currentTrackLocation.x}, ${currentTrackLocation.y}, ${currentTrackLocation.z}, direction; ${currentTrackLocation.direction}`);
		return null;
	}

	// Iterate through the track pieces to find the preceding vehicle.
	const subposition = car.subposition;
	const trackLength = iteratorSegment.getSubpositionLength(subposition, currentTrackLocation.direction);
	let totalDistance = (trackLength - currentProgress);
	for (let i = 0; i < MaxForwardIterations; i++)
	{
		iterator.next();

		const iteratorPosition = iterator.position;
		if (isLocationEqual(iteratorPosition, inFrontTrackLocation))
		{
			// Preceding vehicle found, return total distance between the two.
			return (totalDistance + inFrontProgress);
		}

		iteratorSegment = iterator.segment;
		if (iteratorSegment === null)
		{
			// End of track, look no further.
			return null;
		}

		// Add distance of current track piece and check the next one.
		totalDistance += iteratorSegment.getSubpositionLength(subposition, iteratorPosition.direction);
	}
	return null;
}


function getTrackIteratorAtLocation(trackLocation: CoordsXYZD): TrackIterator | null
{
	const currentTrackIndex = getIndexForTrackElementAt(trackLocation);
	if (currentTrackIndex === null)
	{
		console.log(`Could not find track for car at position; ${trackLocation.x}, ${trackLocation.y}, ${trackLocation.z}, direction; ${trackLocation.direction}`);
		return null;
	}

	const iterator = map.getTrackIterator(trackLocation, currentTrackIndex);
	if (!iterator)
	{
		console.log(`Could not start track iterator for car at position; ${trackLocation.x}, ${trackLocation.y}, ${trackLocation.z}, direction; ${trackLocation.direction}, index; ${currentTrackIndex}`);
		return null;
	}
	return iterator;
}


function getIndexForTrackElementAt(coords: CoordsXYZD): number | null
{
	const tile = map.getTile(coords.x / 32, coords.y / 32);
	const allElements = tile.elements, len = allElements.length;

	for (let i = 0; i < len; i++)
	{
		const element = tile.elements[i];
		if (element.type === "track"
			&& element.baseZ === coords.z
			&& element.direction === coords.direction)
		{
			return i;
		}
	}
	return null;
}


function isLocationEqual(left: CoordsXYZD, right: CoordsXYZD): boolean
{
	return left.x === right.x
		&& left.y === right.y
		&& left.z === right.z
		&& left.direction === right.direction;
}
```

Comment (IntelOrca) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:149]: Cool, that looks fine

**Review by IntelOrca — APPROVED** [review]
